solarextrapolation/classes.py

```diff
@@ -0,0 +1,333 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Fri Jun 05 12:49:03 2015
+
+@author: Alex
+
+This module was developed with funding provided by ESA Summer of Code in Space
+(2015).
+"""
+
+import sunpy.map
+import numpy as np
+import pickle
+import time
+#import astropy.units as u
+
+class Preprocessors(object):
+    """
+    A common class for all 2D pre-processing routines, tools used to pre-process
+    the 2D sunpy map data for use in extrapolations.
+    Usage can include basic filters for noise/contrast or algorythms to
+    compensate for extrapolator assumptions, such as the force-free assumption
+    that is assumed in many extrapolations, but isn't true in the photosphere
+    where magnetogram observations are generally taken.
+    """
+    def __init__(self, map_data, **kwargs):
+        """Method for creating a preprocessor object, using a sunpy map.
+        """
+        # Add some type checking, we want a map object, check for .unit attribute.
+        self.map_input = map_data
+        self.routine = kwargs.get('preprocessor_routine', type(self))
+        self.meta = { 'original_map_header': self.map_input.meta, 'preprocessor_routine': self.routine }
+        self.filepath = kwargs.get('filepath', None)
+
+    def _preprocessor(self):
+        """
+        Method running the and returning a sunpy map.
+        For tracability this should add entries into the metadata that
+        include any parameters used for the given run.
+        """
+        map_output = sunpy.map.Map(self.map_input.data, self.meta)
+        return map_output
+
+    def preprocess(self):
+        
+        """
+        Method to be called to run the preprocessor.
+        Times the process and saves output where applicable.
+        """
+        tim_start = time.time()
+        map_output = self._preprocessor()
+        tim_duration = time.time() - tim_start
+
+        map_output.meta['preprocessor_start_time'] = tim_start
+        map_output.meta['preprocessor_duration'] = tim_duration
+
+        if self.filepath:
+            map_output.save(self.filepath)
+        return map_output
+    
+
+
+class Extrapolators(object):
+    """
+    Common class for all 3D vector field extrapolation routines.
+    Each routine, created by building a subclass, will have wildly varying
+    capabilities and input arguments so this have been left intentionally
+    minimal.
+    The primary method to override is extrapolation(), the primary method to
+    call is extrapolate() which will both call extrapolation() and save the
+    result if a filepath argument is given.
```

Comment:
You should document the parameters for the `__init__` function in the main docstring for the class.

i.e.


```python

Parameters
----------

map_boundary_data : `sunpy.map.GenericMap`
    the sunpy map containing the boundary data.

...

```